src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,215 @@
+     * @param label A reference to the mutable char span which will be mutated to receive the label on success. Use
+     * CopyCharSpanToMutableCharSpan to copy into the MutableCharSpan.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label) = 0;
+
+    /**
+     * Get the mode value of the Nth mode in the list of modes.
+     * @param modeIndex The index of the mode to be returned. It is assumed that modes are indexable from 0 and with no gaps.
+     * @param value a reference to the uint8_t variable that is to contain the mode value.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR GetModeValueByIndex(uint8_t modeIndex, uint8_t & value) = 0;
+
+    /**
+     * Get the mode tags of the Nth mode in the list of modes.
+     * The caller must make sure the List points to an existing buffer of sufficient size to hold the spec-required number
+     * of tags, and the size of the List is the size of the buffer.
+     *
+     * The implementation must place its desired ModeTagStructType instances in that buffer and call tags.reduce_size
+     * on the list to indicate how many entries were initialized.
+     * @param modeIndex The index of the mode to be returned. It is assumed that modes are indexable from 0 and with no gaps.
+     * @param tags a reference to an existing and initialised buffer that is to contain the mode tags. std::copy can be used
+     * to copy into the buffer.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR GetModeTagsByIndex(uint8_t modeIndex,
+                                          DataModel::List<chip::app::Clusters::detail::Structs::ModeTagStruct::Type> & modeTags) = 0;
+
+    /**
+     * When a ChangeToMode command is received, if the NewMode value is a supported made, this function is called to 1) decide if
+     * we should go ahead with transitioning to this mode and 2) formulate the ChangeToModeResponse that will be sent back to the
+     * client. If this function returns a response.status of StatusCode::kSuccess, the change request is accepted
+     * and the CurrentMode is set to the NewMode. Else, the CurrentMode is left untouched. The response is sent as a
+     * ChangeToModeResponse command.
+     *
+     * This function is to be overridden by a user implemented function that makes this decision based on the application logic.
+     * @param NewMode The new made that the device is requested to transition to.
+     * @param response A reference to a response that will be sent to the client. The contents of which con be modified by the
+     * application.
+     *
+     */
+    virtual void HandleChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response) = 0;
```

Comment:
How does the callee set the StatusText usefully?  Do we not plan to support that?

Reply:
Ah, indeed.  I don't know what I was thinking when I wrote that comment....